### PR TITLE
Add batched_grad parameter to `autograd.grad`

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65566
* __->__ #65564
* #65565

- wrap the call into engine with vmap if `batched_grad` is `True`
- improves the comment on the call to engine (somewhat addressing https://github.com/pytorch/pytorch/issues/41659)
- borrows the message from functional.jacobian's vectorized argument concerning usage of the vmap feature
- adds basic test (further testing is done when we replace the usage in vectorized jacobian computation)

TODO:
 - create an issue tracking this

Differential Revision: [D31236259](https://our.internmc.facebook.com/intern/diff/D31236259)

### CODE REVIEW

File: torch/autograd/__init__.py [hunk 1]
```diff
@@ -147,13 +150,13 @@ def backward(
         tuple(inputs) if inputs is not None else tuple()
 
     grad_tensors_ = _tensor_or_tensors_to_tuple(grad_tensors, len(tensors))
-    grad_tensors_ = _make_grads(tensors, grad_tensors_)
+    grad_tensors_ = _make_grads(tensors, grad_tensors_, batched_grads=False)
     if retain_graph is None:
         retain_graph = create_graph
 
     Variable._execution_engine.run_backward(
```
Comment (vadimkantorov) [inline torch/autograd/__init__.py:157]: Still think that the best would be to just collapse these three lines into a single one for fixing https://github.com/pytorch/pytorch/issues/41659 (even if this needs suppressing the syntax linter for this particualr case)

File: torch/autograd/__init__.py [hunk 2]
```diff
@@ -21,16 +21,19 @@
...
 
-def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor]) -> Tuple[_OptionalTensor, ...]:
+def _make_grads(outputs: Sequence[torch.Tensor], grads: Sequence[_OptionalTensor],
+                batched_grads: bool) -> Tuple[_OptionalTensor, ...]:
     new_grads: List[_OptionalTensor] = []
     for out, grad in zip(outputs, grads):
         if isinstance(grad, torch.Tensor):
-            if not out.shape == grad.shape:
+            grad_shape = grad.shape if not batched_grads else grad.shape[1:]
+            if not out.shape == grad_shape:
```
Comment (albanD) [inline torch/autograd/__init__.py:36]: Should the error message be adapted below?

File: torch/autograd/__init__.py [hunk 3]
```diff
@@ -147,13 +150,13 @@ def backward(
...
 
     grad_tensors_ = _tensor_or_tensors_to_tuple(grad_tensors, len(tensors))
-    grad_tensors_ = _make_grads(tensors, grad_tensors_)
+    grad_tensors_ = _make_grads(tensors, grad_tensors_, batched_grads=False)
     if retain_graph is None:
         retain_graph = create_graph
 
     Variable._execution_engine.run_backward(
         tensors, grad_tensors_, retain_graph, create_graph, inputs,
-        allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
```
Comment (albanD) [inline torch/autograd/__init__.py:156]: nit if this comment is for the error message, maybe mention that this is going to c++?

File: torch/autograd/__init__.py [hunk 4]
```diff
@@ -203,6 +207,18 @@ def grad(
         allow_unused (bool, optional): If ``False``, specifying inputs that were not
             used when computing outputs (and therefore their grad is always zero)
             is an error. Defaults to ``False``.
+        batched_grads (bool, optional): If ``True``, the first dimension of each
```
Comment (albanD) [inline torch/autograd/__init__.py:210]: For boolean args, I prefer `is_grads_batched` to make it clear from the name that it is boolean and does not contain actual gradients

File: torch/autograd/__init__.py [hunk 5]
```diff
@@ -203,6 +207,18 @@ def grad(
...
+            tensor in ``grad_outputs`` will be interpreted as the batch dimension.
+            Instead of computing a single vector-Jacobian product, we compute a
+            batch of vector-Jacobian products for each "vector" in the batch.
+            We use the vmap prototype feature as the backend to vectorize calls
+            to the autograd engine so that this computation can be performed in a
+            single call. This should lead to performance improvements when compared
+            to manually looping and performing backward multiple times. Note that
+            due to this feature being incomplete, there may be performance
```
Comment (albanD) [inline torch/autograd/__init__.py:218]: incomplete -> experiemental?

File: torch/autograd/__init__.py [hunk 6]
```diff
@@ -203,6 +207,18 @@ def grad(
...
+            cliffs. Please use `torch._C._debug_only_display_vmap_fallback_warnings(True)`
+            to show any performance warnings and file us issues if warnings exist
```
Comment (albanD) [inline torch/autograd/__init__.py:220]: file us -> file on github

**Review by albanD — COMMENTED** [review]
Looks pretty good.
test though?

File: torch/autograd/__init__.py (hunk 3 above)
Comment (soulitzer) [reply torch/autograd/__init__.py:156]: Ended up going with Vadim's suggestion, but also added a comment to point this out

File: torch/autograd/__init__.py (hunk 1 above)
Comment (soulitzer) [reply torch/autograd/__init__.py]: Hmm, I'm okay with this

File: torch/autograd/__init__.py (hunk 3 above)
Comment (albanD) [reply torch/autograd/__init__.py:156]: I actually think that the comment is better than the one liner.
Pointing the users to `Variable._execution_engine` is going to lead them nowhere.

Comment (soulitzer) [reply torch/autograd/__init__.py:156]: The comment does seem to be more useful, but I guess for the more experienced user, it is more about aesthetics than usefulness lol.

If we only care about newer versions of python, the multi-line issue correctly displays the first line on Python 3.8 it seems https://bugs.python.org/issue12458.

I'm okay with just doing:
```
  File "/home/local/pytorch/torch/autograd/__init__.py", line 267, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/home/local/pytorch/torch/autograd/function.py", line 199, in apply
    return user_fn(self, *args)
  File "/home/local/test.py", line 10, in backward
    assert False
AssertionError
```

Comment (soulitzer) [reply torch/autograd/__init__.py:156]: From offline discussion: we should duplicate the comment on both lines, and add comment above

**Review by albanD — APPROVED** [review]
LGTM only duplicating the comment on the engine call to make sure it displays for all python versions.

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

File: test/test_autograd.py [hunk 7]
```diff
@@ -719,6 +719,30 @@ def backward(ctx, x):
...
+
+        out = x.clone()  # Size([2, 2])
+        batched_grad = torch.arange(3).expand(2, 2, 3).transpose(0, 2)  # Size([3, 2, 2])
+        grad, = torch.autograd.grad(out, (x,), (batched_grad,), is_grads_batched=True)
+        self.assertEqual(grad, torch.arange(3).expand(2, 2, 3).transpose(0, 2).to(dtype=grad.dtype))
+
+        # Detect shape mismatch
+        grad_out = torch.ones(2, 2)
+        with self.assertRaisesRegex(RuntimeError, "If `is_grads_batched=True`, we interpret the first"):
+            torch.autograd.grad(outputs=out, grad_outputs=(grad_out,), inputs=(x,), is_grads_batched=True)
```
Comment (zou3519) [inline test/test_autograd.py:733]: If vmap ends up in pytorch core, would we still want to have the `is_grads_batched=True` API? Because you can do this with by using vmap:
```
vmap(torch.autograd.grad, (None, None, 0))(out, x, grad_out)
```

Comment (soulitzer) [reply test/test_autograd.py:733]: One reason for having both is that some users might be more comfortable with imperative APIs over a functional one,
though I can see that its not super ideal to have two ways of doing the same thing.

File: torch/autograd/__init__.py [hunk 8]
```diff
@@ -21,21 +21,35 @@
...
-                                   + str(grad.shape) + " and output["
-                                   + str(outputs.index(out)) + "] has a shape of "
-                                   + str(out.shape) + ".")
+            grad_shape = grad.shape if not is_grads_batched else grad.shape[1:]
+            if not out.shape == grad_shape:
+                if is_grads_batched:
+                    raise RuntimeError("If `is_grads_batched=True`, we interpret the first "
+                                       "dimension of each grad_output as the batch dimension. "
+                                       "The sizes of the remaining dimensions are expected to match "
+                                       "the shape of corresponding output, but a mismatch "
```
Comment (zou3519) [inline torch/autograd/__init__.py:41]: Maybe suggest using vmap if the user actually wanted only some of their grad_output to be batched

File: test/test_autograd.py (hunk 7 above)
Comment (zou3519) [reply test/test_autograd.py:733]: Paging @albanD for discussion. Either way I'm fine with this going in as-is (we can always revisit the messaging later...)

Comment (albanD) [reply test/test_autograd.py:733]: I think this is a good question. I think it is ok to have both, especially because this one is not necessarily related to vmap. And I can see people using this one without even knowing what vmap is.

**Review by zou3519 — APPROVED** [review]

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31236259).
